landlab/components/profiler/trickle_down_profiler.py

```diff
@@ -0,0 +1,392 @@
+    **Required Software Citation(s) Specific to this Component**
+
+    None Listed
+
+    **Additional References**
+
+    None Listed
+
+    """
+
+    _name = "TrickleDownProfiler"
+
+    _unit_agnostic = True
+
+    _info = {
+        "flow__link_to_receiver_node": {
+            "dtype": int,
+            "intent": "in",
+            "optional": False,
+            "units": "-",
+            "mapping": "node",
+            "doc": "ID of link downstream of each node, which carries the discharge",
+        },
+        "flow__receiver_node": {
+            "dtype": int,
+            "intent": "in",
+            "optional": False,
+            "units": "-",
+            "mapping": "node",
+            "doc": "Node array of receivers (node that receives flow from current node)",
+        },
+    }
+
+    def __init__(
+        self, grid, starting_nodes=None, cmap="viridis",
+    ):
+        """
+        Parameters
+        ----------
+        grid : Landlab Model Grid instance, required
+        starting_nodes : iterable
+        cmap : str
+            A valid matplotlib cmap string. Default is "viridis".
+
+        """
+        super(TrickleDownProfiler, self).__init__(grid)
+
+        self._cmap = plt.get_cmap(cmap)
+
+        self._flow_receiver = grid.at_node["flow__receiver_node"]
+        self._starting_nodes = starting_nodes
+
+    @property
+    def data_structure(self):
+        """OrderedDict defining the trickle down channel network.
+
+        The node IDs and distances upstream of the channel network are stored
+        in ``data_structure``. It is a dictionary with keys of the outlet node
+        ID.
+
+        For each starting node, the value in the ``data_structure`` is
+        itself a dictionary with keys that are a segment ID tuple of the
+        ``(dowstream, upstream)`` nodes IDs of each channel segment.
+
+        The value associated with the segment ID tuple
+        ``(dowstream, upstream)`` is itself a dictionary. It has three
+        key-value pairs. First, ``"ids"`` contains a list of the segment node
+        IDs ordered from downstream to upstream. It includes the endpoints.
+        Second, ``"distances"`` contains a list of distances upstream that
+        mirrors the list in ``"ids"``. Finally, ``"color"`` is an RGBA tuple
+        indicating the color for the segment.
+        """
+        return self._data_struct
+
+    def _create_profile_structure(self):
+        """Create the profile_IDs data structure for channel network.
+
+        The bound attribute self._profile structure is the channel segment
+        datastructure. Profile structure is a list of length
+        starting_nodes. Each element of profile_structure is itself a
+        list of length number of stream segments that drain to each of the
+        starting nodes. Each stream segment list contains the node ids of a
+        stream segment from downstream to upstream.
+        """
+        self._data_struct = OrderedDict()
+
```

Comment:
These loops are probably plenty fast going downstream. If some optimization is desired (and given me doing a too granular of a review), I understand performance can be improved by [eliminating dots and making variables local before the loop](https://wiki.python.org/moin/PythonSpeed/PerformanceTips#Avoiding_dots...), e.g.

in for loop, before while:
```python
channel_segment = []
append = channel_segment.append
```

and maybe you have experience if this even helps minutely
before for loop:
```python
flow_receiver = self._flow_receiver
data_struct = self._data_struct
```

Reply:
neat! I don't have the bandwidth to test if this makes a difference right now but will file it away and try it out at somepoint.